### DISCUSSION TITLE
Disabling Nagle's algorithm

### DISCUSSION THREAD

**Discussion by haljin** [nats-io/nats-server#4267] [Q&A] [answered]
Hi,

I was wondering if it is possible to disable Nagle's algorithm on NATS TCP connections or maybe if that is already done. That applies to both client connections and the intra-cluster connections.

We have a soft real-time application that today is using RabbitMQ but we would like to switch over to using NATS instead as it seems to handle rolling updates in K8s a lot better. Pretty much every TCP connection in our system today is running with Nagle's algorithm disabled (TCP_NODELAY). I have noticed a huge improvement in our request/response latency between two services when I set TCP_NODELAY on all the clients (that is both the client and the server that responds to requests via NATS):

![image](https://github.com/nats-io/nats-server/assets/2393962/e9f1925c-78d2-43e8-862b-c1ff85071f06)

If you just ignore the outlying spike, the maximum latency observed dropped from a pretty constant 50ms to ~15ms. The average has also improved considerably:

![image](https://github.com/nats-io/nats-server/assets/2393962/a98acec3-c48f-4d40-93dc-c1ac7da1fd25)

However we are running a 3-node NATS cluster, so I imagine some of the messaging must pass between the nodes in the cluster (since there are multiple clients and multiple servers). I'd like to be able to set TCP_NODELAY on those connections as well if possible.

**Comment by jgaskins** [comment, marked as answer]
Golang [disables Nagle's algorithm by default](https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/net/tcpsock.go;l=239) (a choice [not everyone agrees with](https://withinboredom.info/2022/12/29/golang-is-evil-on-shitty-networks/)) and the NATS server [doesn't call `SetNoDelay`](https://github.com/search?q=repo%3Anats-io%2Fnats-server+SetNoDelay&type=code) anywhere in the codebase to override that. So you're good.
